data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -0,0 +1,152 @@
+        this.retryCount = retryCount;
+        this.backOffTimeBase = backOffTimeBase;
+
+        this.bufferedEventHandles = new ArrayList<>();
+
+        bufferLock = new ReentrantLock();
+        sinkStopWatch = new SinkStopWatch();
+        asyncExecutor = newCachedThreadPool();
+    }
+
+    /**
+     * Function handles the packaging of events into log events before sending a bulk request to CloudWatchLogs.
+     * Implements simple conditional buffer. (Sends once batch size, request size in bytes, or time limit is reached)
+     * @param logs - Collection of Record events which hold log data.
+     */
+    public void processLogEvents(final Collection<Record<Event>> logs) {
+        sinkStopWatch.startIfNotRunning();
+        for (Record<Event> log: logs) {
+            String logString = log.getData().toJsonString();
+            int logLength = logString.length();
+
+            if (cloudWatchLogsLimits.isGreaterThanMaxEventSize(logLength)) {
+                LOG.warn("Event blocked due to Max Size restriction! {Event Size: {} bytes}", (logLength + CloudWatchLogsLimits.APPROXIMATE_LOG_EVENT_OVERHEAD_SIZE));
+                continue;
+            }
+
+            long time = sinkStopWatch.getStopWatchTimeSeconds();
+
+            bufferLock.lock();
+
+            int bufferSize = buffer.getBufferSize();
+            int bufferEventCount = buffer.getEventCount();
+            int bufferEventCountWithEvent = bufferEventCount + 1;
+            int bufferSizeWithAddedEvent = bufferSize + logLength;
+
+            if ((cloudWatchLogsLimits.isGreaterThanLimitReached(time, bufferSizeWithAddedEvent, bufferEventCountWithEvent) && (bufferEventCount > 0))) {
+                stageLogEvents();
+                addToBuffer(log, logString);
+            } else if (cloudWatchLogsLimits.isEqualToLimitReached(bufferSizeWithAddedEvent, bufferEventCountWithEvent)) {
+                addToBuffer(log, logString);
+                stageLogEvents();
+            } else {
+                addToBuffer(log, logString);
+            }
+
+            bufferLock.unlock();
+        }
+    }
+
+    private void stageLogEvents() {
+        sinkStopWatch.stopAndResetStopWatch();
+
+        List<byte[]> eventMessageClone = buffer.getBufferedData();
+
+        ThreadTaskEvents dataToPush = new ThreadTaskEvents(eventMessageClone, bufferedEventHandles);
+        taskQueue.add(dataToPush);
+
+        buffer.resetBuffer();
+        bufferedEventHandles = new ArrayList<>();
+
+        CloudWatchLogsDispatcher newTaskDispatcher = CloudWatchLogsDispatcher.builder()
+                .taskQueue(taskQueue)
+                .cloudWatchLogsClient(cloudWatchLogsClient)
+                .cloudWatchLogsMetrics(cloudWatchLogsMetrics)
+                .logGroup(logGroup)
+                .logStream(logStream)
+                .retryCount(retryCount)
+                .backOffTimeBase(backOffTimeBase)
+                .build();
+
+        asyncExecutor.execute(newTaskDispatcher);
```

Comment:
I dont think creating an entire `CloudWatchLogsDispatcher` is the way to go. This class is fairly big and includes a lot of stuff we can run on the same thread and stuff we don't need to re-create all the time.

Reply:
Agreed, I will try to minimize the the requirements for a log publisher. This way we can have more lightweight objects running on each thread.